radix-engine-tests/tests/db/substate_database_overlay.rs

```diff
@@ -484,30 +484,59 @@ fn run_scenarios(
         .build();
 
     let mut next_nonce: u32 = 0;
-    for scenario_builder in get_builder_for_every_scenario() {
-        let epoch = ledger1.get_current_epoch();
-        let mut scenario = scenario_builder(ScenarioCore::new(network.clone(), epoch, next_nonce));
-        let mut previous = None;
-        loop {
-            let next = scenario
-                .next(previous.as_ref())
-                .map_err(|err| err.into_full(&scenario))
-                .unwrap();
-            match next {
-                NextAction::Transaction(next) => {
-                    let receipt1 = ledger1.execute_notarized_transaction(&next.raw_transaction);
-                    let receipt2 = ledger2.execute_notarized_transaction(&next.raw_transaction);
-
-                    check_callback(
-                        (ledger1.substate_db(), &receipt1),
-                        (ledger2.substate_db(), &receipt2),
-                    );
-
-                    previous = Some(receipt1);
-                }
-                NextAction::Completed(end_state) => {
-                    next_nonce = end_state.next_unused_nonce;
-                    break;
+    let mut scenario_builders = scenario_builders();
+    for requirement in ScenarioRequirements::all() {
+        if let ScenarioRequirements::ProtocolUpdateUpTo(protocol_update) = requirement {
+            protocol_update
+                .generate_state_updates(ledger1.substate_db_mut(), &network)
```

Comment:
Apologies, name here is not clear, this part makes the assumption that a previous requirement seen by the loop was the anemone protocol update and that anemone was applied at that point, this is true since we iterate over `ScenarioRequirements::all()` in this loop. Seems to me like a name change might be better here, or perhaps even switching to the vectors based approach you left in the other comment.

Reply:
This has been refactored to make it clear that each protocol update is applied only once and that is when it is encountered.